docs: improve twister docs (#39)
IRNAS/irnas-zephyr-template: doc/development/twister.md

# Twister

Twister is a script that scans for the set of unit test applications in the git repository and
attempts to execute them. It can build the applications, run them (if possible in the current build
environment), check if they every successful and generate a report.

> In this project the Twister is used for three distinct purposes:

Author: Thank you for this, my knowledge of `the` usage is quite bad, I need to work on it. I fixed this in the whole file.

  do any tag filtering. Multiple invocations are treated as a logical 'or' relationship.
- `-b, --build-only` - Only build the code, do not attempt to run the code on targets.
- `-f, --only-failed` - Run only those tests that failed the previous twister run invocation.
- `-y, --dry-run` - Create the filtered list of test cases, but don't actually run them. Useful if
  you're just interested in the test plan generated for every run and saved in the specified output
  directory (`testplan.json`).
- `--overflow-as-errors` - Treat FLASH/RAM/SRAM overflows as errors. **This one is important. If not
  used the test that overflows is considered to be skipped and not that it has failed.**
- `-v, --verbose` - Emit build logs, call multiple times to increase verbosity.

## `twister-out` folder

With every invocation of the `east twister` a `twister-out` folder is created. It contains all build
folders of the created projects, as well as the report files.

Keep in mind, if you keep running `east twister` command and not deleting the `twister-out` folder,
then the command will rename it to `twister-out.X` (Where `X` is a unique ascending number, starting

> with 1) before running the full build.

Reviewer: ```suggestion
with 1) before running the full build. Thus, `twister-out` is always the result of the latest twister invocation.
```

## Cookbook style examples

### Running tests on the Host PC

> You want to build and run all tests under `tests/` directory on the Host PC.
> 
> `testcase.yaml` file in the all projects under `tests/` directory should look like this:

Reviewer: Reading this it kind of sounds like all tests must have `platform_allow: native_sim`, but this is not true.

The command will run all tests that have `platform_allow: native_sim`.
You should reword this.

```suggestion
You want to build and run all tests under the `tests/` directory meant for native testing.

`testcase.yaml` file in all the applicable projects under the `tests/` directory should contain this:
```

Author: I have kept the Host PC part, reworded rest.

```yaml
tests:
  tests.test_name.native:
    platform_allow: native_sim
    harness: ztest
```

**Remember**: Test Scenarios must be unique across the whole repository.

From the project's root directory run:

```shell
east twister -p native_sim -T tests
```

### Running tests on the target

You want to build and run all tests under `tests/` directory on the target

`testcase.yaml` file in the all projects under `tests/` directory should look like this:

```yaml
tests:
  tests.test_name.target:
    platform_allow: nrf52dk/nrf52832
    harness: ztest

> ```

Reviewer: Same as above, it sounds like I can either have tests for target only or for native only.
But I can have tests for both, and some tests for only one of them.

From the project's root directory run:

```shell
east twister -T tests -p nrf52dk/nrf52832 --device-testing --device-serial /dev/ttyACM0
```

### Filtering by tags

You want to only build all projects in the repository that have a tag `some_tag`.

`testcase.yaml` file should then look like this:

```yaml
tests:
  samples.bitshifter.with_tag:
    platform_allow: nrf52dk/nrf52832
    tag: some_tag
  samples.bitshifter.without_tag:
    platform_allow: nrf52dk/nrf52832
```

From the project's root directory run:

```shell

> east twister -T . -t some_tag

Reviewer: ```suggestion
east twister -T . -t some_tag  --build-only
```

Since you say "only build", did you actually want this?

Author: Ah, idea was that this sample demonstrates only a single thing (filtering by tag), but now I see that won't work due to the missing platform flag. I expanded this:

`````
You want to only build (and not run) all projects in the repository that have a tag `some_tag` on
`nrf52dk/nrf52832` target.

`testcase.yaml` file should then look like this:

```yaml
tests:
  samples.bitshifter.with_tag:
    platform_allow: nrf52dk/nrf52832
    tag: some_tag
  samples.bitshifter.without_tag:
    platform_allow: nrf52dk/nrf52832
```

From the project's root directory run:

```shell
east twister -T . -t some_tag -p nrf52dk/nrf52832 --build-only
```
`````

```

### platform_allow acts as a list and not as a filter

You want `platform_allow` to act as a list of platforms that Twister should consider and not as a
filter, as you want to avoid the [scope presumption] behavior. You don't want to specify the
`--platform` option as this can be hard to maintain.

[scope presumption]:
  https://docs.zephyrproject.org/latest/develop/test/twister.html#selecting-platform-scope

`testcase.yaml` file should then look like this:

```yaml
tests:
  app.rtt:
    platform_allow: nrf52dk/nrf52832 nrf52840dk/nrf52840
  app.debug:
    platform_allow: nrf52dk/nrf52832 nrf52840dk/nrf52840 native_sim
```

Extra `twister_config.yaml` file is needed:

```yaml
platforms:
  override_default_platforms: true
```

From the project's root directory run:

```shell
east twister -T . --test-config twister_config.yaml
```

## Coverage report

Twister supports creating coverage reports from test runs with `gcovr` tool. Adding extra flags to
the command invocation creates a html file (among others) which shows code coverage for each file.

```shell
east twister -T tests -p native_sim --coverage --coverage-tool gcovr
```

> Note that you might need to change `--coverage-basedir` or add new ones for your project.

Reviewer: <!--__GRAPHITE_HTML_TAG_START__--><p class='graphite__hidden'><i>[Re: line +289]</i></p><!--__GRAPHITE_HTML_TAG_END__-->
Maybe the make section should be in its own file?
<!--__GRAPHITE_HTML_TAG_START__--><p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.dev/github/pr/IRNAS/irnas-zephyr-template/39?utm_source=unchanged-line-comment">Graphite</a>.</p><!--__GRAPHITE_HTML_TAG_END__-->

Author: Let's keep it in here for now, we can revisit this later.
